mono/utils/mono-threads.c

```diff
@@ -791,6 +791,82 @@ mono_thread_info_set_flags (MonoThreadInfoFlags flags)
 		threads_callbacks.thread_flags_changed (old, flags);
 }
 
+struct GSList {
+  gpointer data;
+  GSList *next;
+};
+
+#define MONO_END_INIT_CB GINT_TO_POINTER(-1)
+static GSList *init_callbacks;
+
+void
+mono_thread_info_wait_inited (void)
+{
+	MonoSemType cb;
+	mono_os_sem_init (&cb, 0);
+	gpointer old = init_callbacks;
+
+	GSList wait_request;
+	wait_request.data = &cb;
+	wait_request.next = old;
+
+	while (mono_threads_inited != TRUE) {
+		gpointer old_read = mono_atomic_cas_ptr ((gpointer *) &init_callbacks, &wait_request, old);
+
+		// Queued up waiter, need to be unstuck
+		if (old_read == old) {
+			break;
+		} else if (old_read == GINT_TO_POINTER (MONO_END_INIT_CB)) {
+			// Is inited
+			return; 
+		} else {
+			// We raced with another writer
+			wait_request.next = (GSList *) old_read;
+			old = old_read;
+		}
+	}
+
+	while (mono_threads_inited != TRUE) {
+		gboolean timedout = mono_os_sem_timedwait (&cb, 1000, MONO_SEM_FLAGS_NONE) == MONO_SEM_TIMEDWAIT_RET_TIMEDOUT;
+		if (!timedout)
+			break;
+	}
+
+	g_assert (mono_threads_inited);
+	return;
+}
+
+static void
+mono_thread_info_set_inited (void)
+{
+	mono_threads_inited = TRUE;
+	mono_memory_barrier ();
+
+	GSList *old = init_callbacks;
+
+	while (TRUE) {
+		gpointer old_read = mono_atomic_cas_ptr ((gpointer *) &init_callbacks, MONO_END_INIT_CB, (gpointer) old);
+		if (old == old_read)
+			break;
+		else
+			old = old_read;
+	}
+	if (old == MONO_END_INIT_CB) {
+		g_warning ("Global threads inited twice");
```

Comment:
We should probably `g_error` here as multiple calls to `mono_thread_info_init` would have pretty disastrous results.

Reply:
Thanks, will fix. And it does. I was doing it accidentally for a while and the crashes were weird TLS variable issues that all looked like totally unrelated things